subprojects/core/src/main/java/org/gradle/internal/vfs/RoutingVirtualFileSystem.java

```diff
@@ -83,8 +83,10 @@ public void update(Iterable<String> locations, Runnable action) {
 
     @Override
     public void invalidateAll() {
-        gradleUserHomeVirtualFileSystem.invalidateAll();
         buildScopedVirtualFileSystem.invalidateAll();
+        if (vfsRetained.getAsBoolean()) {
+            gradleUserHomeVirtualFileSystem.invalidateAll();
+        }
```

Comment:
Do we need to invalidate `buildScopedVirtualFileSystem` if `vfsRetained` is on?

Reply:
I changed it now that we only invalidate one of the two VFSs.